File: rust/crates/planet-core/src/cubesphere.rs
//! Cube-sphere parameterisation. See the Python docstring for the convention;
//! in short: right-handed, +Z north pole, +X through (lat 0, lon 0), faces
//! 0..5 = +X -X +Y -Y +Z -Z, tangent-warped (u, v) in [-1, 1]^2.

use std::f64::consts::{FRAC_PI_4, PI};

pub type Vec3 = [f64; 3];

/// (n, tu, tv) per face with `tu x tv == n`.
pub const FACE_AXES: [[Vec3; 3]; 6] = [
    [[1.0, 0.0, 0.0], [0.0, 1.0, 0.0], [0.0, 0.0, 1.0]],
    [[-1.0, 0.0, 0.0], [0.0, -1.0, 0.0], [0.0, 0.0, 1.0]],
    [[0.0, 1.0, 0.0], [-1.0, 0.0, 0.0], [0.0, 0.0, 1.0]],
    [[0.0, -1.0, 0.0], [1.0, 0.0, 0.0], [0.0, 0.0, 1.0]],
    [[0.0, 0.0, 1.0], [1.0, 0.0, 0.0], [0.0, 1.0, 0.0]],
    [[0.0, 0.0, -1.0], [0.0, 1.0, 0.0], [1.0, 0.0, 0.0]],
];

pub const FACE_NAMES: [&str; 6] = ["px", "nx", "py", "ny", "pz", "nz"];

pub fn warp(u: f64) -> f64 {
    (u * FRAC_PI_4).tan()
}

pub fn unwarp(s: f64) -> f64 {
    s.atan() * 4.0 / PI
}

fn dot(a: Vec3, b: Vec3) -> f64 {
    a[0] * b[0] + a[1] * b[1] + a[2] * b[2]
}

pub fn normalize(d: Vec3) -> Vec3 {
    let n = dot(d, d).sqrt();
    [d[0] / n, d[1] / n, d[2] / n]
}

/// Unit direction for face coordinates.
pub fn face_uv_to_dir(face: usize, u: f64, v: f64) -> Vec3 {
    let [n, tu, tv] = FACE_AXES[face];
    let (s, t) = (warp(u), warp(v));
    normalize([
        n[0] + s * tu[0] + t * tv[0],
        n[1] + s * tu[1] + t * tv[1],
        n[2] + s * tu[2] + t * tv[2],
    ])
}

/// Inverse of [`face_uv_to_dir`]; `d` need not be normalised.
pub fn dir_to_face_uv(d: Vec3) -> (usize, f64, f64) {
    let mut axis = 0;
    for k in 1..3 {
        if d[k].abs() > d[axis].abs() {
            axis = k;
        }
    }
    let face = axis * 2 + usize::from(d[axis] < 0.0);
    let [n, tu, tv] = FACE_AXES[face];
    let dn = dot(d, n);
    (face, unwarp(dot(d, tu) / dn), unwarp(dot(d, tv) / dn))
}

/// Degrees; lat in [-90, 90], lon in [-180, 180).
pub fn dir_to_latlon(d: Vec3) -> (f64, f64) {
    let n = dot(d, d).sqrt();
    let lat = (d[2] / n).clamp(-1.0, 1.0).asin().to_degrees();
    let mut lon = d[1].atan2(d[0]).to_degrees();
    if lon >= 180.0 {
        lon -= 360.0;
    }
    (lat, lon)
}

pub fn latlon_to_dir(lat_deg: f64, lon_deg: f64) -> Vec3 {
    let (la, lo) = (lat_deg.to_radians(), lon_deg.to_radians());
    let c = la.cos();
    [c * lo.cos(), c * lo.sin(), la.sin()]
}

/// (u, v) at the centre of pixel (row i, col j) of an n x n face raster.
/// Row 0 is the +v edge, column 0 the -u edge.
pub fn pixel_uv(n: usize, i: usize, j: usize) -> (f64, f64) {
    let u = (j as f64 + 0.5) / n as f64 * 2.0 - 1.0;
    let v = 1.0 - (i as f64 + 0.5) / n as f64 * 2.0;
    (u, v)
}

/// Nearest pixel (row, col) for face coordinates; clamps at the edges.
pub fn uv_to_pixel(n: usize, u: f64, v: f64) -> (usize, usize) {
    let max = (n - 1) as f64;
    let j = ((u + 1.0) / 2.0 * n as f64 - 0.5).round().clamp(0.0, max) as usize;
    let i = ((1.0 - v) / 2.0 * n as f64 - 0.5).round().clamp(0.0, max) as usize;
    (i, j)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn axes_are_right_handed() {
        for [n, tu, tv] in FACE_AXES {
            let c = [
                tu[1] * tv[2] - tu[2] * tv[1],
                tu[2] * tv[0] - tu[0] * tv[2],
                tu[0] * tv[1] - tu[1] * tv[0],
            ];
            assert_eq!(c, n);
        }
    }

    #[test]
    fn pixel_round_trip() {
        for n in [1usize, 8, 4096] {
            for (i, j) in [(0, 0), (n - 1, n - 1), (n / 2, n / 3)] {
                let (u, v) = pixel_uv(n, i, j);
                assert_eq!(uv_to_pixel(n, u, v), (i, j));
            }
        }
    }
}

File: src/planetcreator/cubesphere.py
"""Cube-sphere parameterisation shared by the Python pipeline and the engine.

Conventions (the Rust/Godot side must match these exactly):

* Right-handed axes. +Z is the north pole, +X passes through (lat 0, lon 0),
  +Y through (lat 0, lon 90 E).
* Six faces indexed 0..5 = +X, -X, +Y, -Y, +Z, -Z. Each face has a normal
  ``n`` and tangents ``tu``, ``tv`` with ``tu x tv = n``.
* Face coordinates ``(u, v)`` in [-1, 1]^2. They are tangent-warped
  (``s = tan(u * pi/4)``) before projection, which makes pixel area nearly
  uniform across the face (the S2 "tan" projection).
* A face raster of ``N x N`` pixels has pixel ``(row i, col j)`` centred at
  ``u = (j + 0.5) / N * 2 - 1`` and ``v = 1 - (i + 0.5) / N * 2``:
  row 0 is the +v edge, column 0 the -u edge.
"""

from __future__ import annotations

import numpy as np

# (n, tu, tv) per face; tu x tv == n for every face.
FACE_AXES = np.array(
    [
        [[1, 0, 0], [0, 1, 0], [0, 0, 1]],  # +X
        [[-1, 0, 0], [0, -1, 0], [0, 0, 1]],  # -X
        [[0, 1, 0], [-1, 0, 0], [0, 0, 1]],  # +Y
        [[0, -1, 0], [1, 0, 0], [0, 0, 1]],  # -Y
        [[0, 0, 1], [1, 0, 0], [0, 1, 0]],  # +Z
        [[0, 0, -1], [0, 1, 0], [1, 0, 0]],  # -Z
    ],
    dtype=np.float64,
)
FACE_NAMES = ("px", "nx", "py", "ny", "pz", "nz")


def warp(u: np.ndarray) -> np.ndarray:
    return np.tan(u * (np.pi / 4))


def unwarp(s: np.ndarray) -> np.ndarray:
    return np.arctan(s) * (4 / np.pi)


def face_uv_to_dir(face: np.ndarray, u: np.ndarray, v: np.ndarray) -> np.ndarray:
    """Unit direction vectors, shape (..., 3)."""
    face = np.asarray(face)
    ax = FACE_AXES[face]  # (..., 3, 3)
    s, t = warp(np.asarray(u, dtype=np.float64)), warp(np.asarray(v, dtype=np.float64))
    d = ax[..., 0, :] + s[..., None] * ax[..., 1, :] + t[..., None] * ax[..., 2, :]
    return d / np.linalg.norm(d, axis=-1, keepdims=True)


def dir_to_face_uv(d: np.ndarray) -> tuple[np.ndarray, np.ndarray, np.ndarray]:
    """Inverse of :func:`face_uv_to_dir`. ``d`` need not be normalised."""
    d = np.asarray(d, dtype=np.float64)
    axis = np.argmax(np.abs(d), axis=-1)
    sign = np.take_along_axis(d, axis[..., None], axis=-1)[..., 0] < 0
    face = axis * 2 + sign.astype(np.intp)
    ax = FACE_AXES[face]
    dn = np.einsum("...i,...i->...", d, ax[..., 0, :])
    s = np.einsum("...i,...i->...", d, ax[..., 1, :]) / dn
    t = np.einsum("...i,...i->...", d, ax[..., 2, :]) / dn
    return face, unwarp(s), unwarp(t)


def dir_to_uv_on_face(d: np.ndarray, face: int) -> tuple[np.ndarray, np.ndarray]:
    """(u, v) of directions projected onto a *given* face's tangent plane, unclamped
    (|u| > 1 beyond the face edge). Used to resample a neighbouring face's padded raster."""
    d = np.asarray(d, dtype=np.float64)
    n, tu, tv = FACE_AXES[face]
    dn = d @ n
    return unwarp((d @ tu) / dn), unwarp((d @ tv) / dn)


def dir_to_latlon(d: np.ndarray) -> tuple[np.ndarray, np.ndarray]:
    """Degrees. lat in [-90, 90], lon in [-180, 180)."""
    d = np.asarray(d, dtype=np.float64)
    n = np.linalg.norm(d, axis=-1)
    lat = np.degrees(np.arcsin(np.clip(d[..., 2] / n, -1, 1)))
    lon = np.degrees(np.arctan2(d[..., 1], d[..., 0]))
    lon = np.where(lon >= 180, lon - 360, lon)
    return lat, lon


def latlon_to_dir(lat: np.ndarray, lon: np.ndarray) -> np.ndarray:
    la, lo = np.broadcast_arrays(np.radians(lat), np.radians(lon))
    c = np.cos(la)
    return np.stack([c * np.cos(lo), c * np.sin(lo), np.sin(la)], axis=-1)


def face_pixel_uv(n: int, pad: int = 0) -> tuple[np.ndarray, np.ndarray]:
    """(u, v) at every pixel centre of an n x n face raster, each shape (n, n).

    With ``pad`` the raster is (n + 2 pad)^2 and extends beyond the face (|u| > 1); the
    tangent warp continues smoothly there, so the extra pixels land on neighbouring faces.
    """
    c = (np.arange(-pad, n + pad, dtype=np.float64) + 0.5) / n * 2 - 1
    u, v = np.meshgrid(c, -c)  # rows run from +v (top) to -v
    return u, v


def uv_to_pixel(n: int, u: float, v: float) -> tuple[int, int]:
    """Nearest pixel (row, col) of an n x n face raster for face coordinates; clamps at the edges."""
    j = int(np.clip(np.rint((u + 1.0) / 2.0 * n - 0.5), 0, n - 1))
    i = int(np.clip(np.rint((1.0 - v) / 2.0 * n - 0.5), 0, n - 1))
    return i, j


def face_pixel_latlon(face: int, n: int, pad: int = 0) -> tuple[np.ndarray, np.ndarray]:
    u, v = face_pixel_uv(n, pad)
    return dir_to_latlon(face_uv_to_dir(np.full_like(u, face, dtype=np.intp), u, v))
